File: apps/api/airm/app/quotas/messaging.py
# Copyright © Advanced Micro Devices, Inc., or its affiliates.
#
# SPDX-License-Identifier: MIT

from typing import Literal

from pydantic import AwareDatetime, BaseModel, Field

from ..clusters.messaging import PriorityClass
from ..utilities.enums import GPUVendor


class ClusterQuotaAllocation(BaseModel):
    cpu_milli_cores: int = Field(description="The guaranteed number of CPU milli cores.")
    gpu_count: int = Field(description="The guaranteed number of GPUs.")
    memory_bytes: int = Field(description="The guaranteed memory in bytes.")
    ephemeral_storage_bytes: int = Field(description="The guaranteed ephemeral storage in bytes.")
    quota_name: str = Field(description="The quota name to uniquely identify the quota in the cluster.")
    namespaces: list[str] = Field(description="The list of namespaces to which the quota applies.")


class ClusterQuotasAllocationMessage(BaseModel):
    message_type: Literal["cluster_quotas_allocation"]
    gpu_vendor: GPUVendor | None = Field(None, description="The vendor of the GPU in the cluster")
    quota_allocations: list[ClusterQuotaAllocation] = Field(description="The list of quota allocations to apply.")
    priority_classes: list[PriorityClass] = Field(description="The list of priority classes to configure.")


class ClusterQuotasStatusMessage(BaseModel):
    message_type: Literal["cluster_quotas_status"]
    updated_at: AwareDatetime = Field(description="The timestamp of the update.")
    quota_allocations: list[ClusterQuotaAllocation] = Field(description="The list of quota allocations.")


class ClusterQuotasFailureMessage(BaseModel):
    message_type: Literal["cluster_quotas_failure"]
    reason: str | None = Field(None, description="The reason for the failure.")
    updated_at: AwareDatetime = Field(description="The timestamp of the update.")

File: apps/agent/internal/quotas/messaging.go
// Copyright © Advanced Micro Devices, Inc., or its affiliates.
//
// SPDX-License-Identifier: MIT

package quotas

import (
	"time"

	"github.com/silogen/agent/internal/common"
	"github.com/silogen/agent/internal/messaging"
)

type PriorityClass struct {
	Name     string `json:"name"`
	Priority int32  `json:"priority"`
}

type ClusterQuotaAllocation struct {
	CPUMilliCores         int64    `json:"cpu_milli_cores"`
	GPUCount              int64    `json:"gpu_count"`
	MemoryBytes           int64    `json:"memory_bytes"`
	EphemeralStorageBytes int64    `json:"ephemeral_storage_bytes"`
	QuotaName             string   `json:"quota_name"`
	Namespaces            []string `json:"namespaces"`
}

type ClusterQuotasAllocationMessage struct {
	MessageType      messaging.MessageType    `json:"message_type"`
	GPUVendor        *common.GPUVendor        `json:"gpu_vendor,omitempty"`
	QuotaAllocations []ClusterQuotaAllocation `json:"quota_allocations"`
	PriorityClasses  []PriorityClass          `json:"priority_classes"`
}

type ClusterQuotasStatusMessage struct {
	MessageType      messaging.MessageType    `json:"message_type"`
	UpdatedAt        time.Time                `json:"updated_at"`
	QuotaAllocations []ClusterQuotaAllocation `json:"quota_allocations"`
}

type ClusterQuotaFailureMessage struct {
	MessageType messaging.MessageType `json:"message_type"`
	UpdatedAt   time.Time             `json:"updated_at"`
	Reason      string                `json:"reason"`
}
